File: modules/objc/dispatch/ExtSdkChatThreadManagerWrapper.m
//
//  ExtSdkChatThreadManagerWrapper.m
//  im_flutter_sdk
//
//  Created by asterisk on 2022/5/25.
//

#import "ExtSdkChatThreadManagerWrapper.h"
#import "ExtSdkMethodTypeObjc.h"
#import "ExtSdkToJson.h"

@interface ExtSdkChatThreadManagerWrapper () <EMThreadManagerDelegate>

@end

@implementation ExtSdkChatThreadManagerWrapper

+ (nonnull instancetype)getInstance {
    static ExtSdkChatThreadManagerWrapper *instance = nil;
    static dispatch_once_t predicate;
    dispatch_once(&predicate, ^{
      instance = [[ExtSdkChatThreadManagerWrapper alloc] init];
    });
    return instance;
}

- (void)initSDK {
    [EMClient.sharedClient.threadManager removeDelegate:self];
    [EMClient.sharedClient.threadManager addDelegate:self delegateQueue:nil];
}

- (void)fetchChatThreadDetail:(NSDictionary *)param
               withMethodType:(NSString *)aChannelName
                       result:(nonnull id<ExtSdkCallbackObjc>)result {
    __weak typeof(self) weakSelf = self;
    NSString *threadId = param[@"threadId"];
    [EMClient.sharedClient.threadManager
        getChatThreadFromSever:threadId
                    completion:^(EMChatThread *_Nonnull thread,
                                 EMError *_Nonnull aError) {
                      [weakSelf onResult:result
                          withMethodType:aChannelName
                               withError:aError
                              withParams:[thread toJsonObject]];
                    }];
}

- (void)fetchJoinedChatThreads:(NSDictionary *)param
                withMethodType:(NSString *)aChannelName
                        result:(nonnull id<ExtSdkCallbackObjc>)result {
    __weak typeof(self) weakSelf = self;
    NSString *cursor = param[@"cursor"];
    NSNumber *pageSize = param[@"pageSize"];
    [EMClient.sharedClient.threadManager
        getJoinedChatThreadsFromServerWithCursor:cursor
                                        pageSize:pageSize.intValue
                                      completion:^(
                                          EMCursorResult *_Nonnull aResult,
                                          EMError *_Nonnull aError) {
                                        [weakSelf onResult:result
                                            withMethodType:aChannelName
                                                 withError:aError
                                                withParams:[aResult
                                                               toJsonObject]];
                                      }];
}

- (void)fetchChatThreadsWithParentId:(NSDictionary *)param
                      withMethodType:(NSString *)aChannelName
                              result:(nonnull id<ExtSdkCallbackObjc>)result {
    __weak typeof(self) weakSelf = self;
    NSString *parentId = param[@"parentId"];
    NSString *cursor = param[@"cursor"];
    NSNumber *pageSize = param[@"pageSize"];
    [EMClient.sharedClient.threadManager
        getChatThreadsFromServerWithParentId:parentId
                                      cursor:cursor
                                    pageSize:pageSize.intValue
                                  completion:^(EMCursorResult *_Nonnull aResult,
                                               EMError *_Nonnull aError) {
                                    [weakSelf onResult:result
                                        withMethodType:aChannelName
                                             withError:aError
                                            withParams:[aResult toJsonObject]];
                                  }];
}

- (void)fetchJoinedChatThreadsWithParentId:(NSDictionary *)param
                            withMethodType:(NSString *)aChannelName
                                    result:
                                        (nonnull id<ExtSdkCallbackObjc>)result {
    __weak typeof(self) weakSelf = self;
    NSString *parentId = param[@"parentId"];
    NSString *cursor = param[@"cursor"];
    NSNumber *pageSize = param[@"pageSize"];
    [EMClient.sharedClient.threadManager
        getJoinedChatThreadsFromServerWithParentId:parentId
                                            cursor:cursor
                                          pageSize:pageSize.intValue
                                        completion:^(
                                            EMCursorResult *_Nonnull aResult,
                                            EMError *_Nonnull aError) {
                                          [weakSelf onResult:result
                                              withMethodType:aChannelName
                                                   withError:aError
                                                  withParams:[aResult
                                                                 toJsonObject]];
                                        }];
}

- (void)fetchChatThreadMember:(NSDictionary *)param
               withMethodType:(NSString *)aChannelName
                       result:(nonnull id<ExtSdkCallbackObjc>)result {
    __weak typeof(self) weakSelf = self;
    NSString *threadId = param[@"threadId"];
    NSString *cursor = param[@"cursor"];
    NSNumber *pageSize = param[@"pageSize"];
    [EMClient.sharedClient.threadManager
        getChatThreadMemberListFromServerWithId:threadId
                                         cursor:cursor
                                       pageSize:pageSize.intValue
                                     completion:^(
                                         EMCursorResult *_Nonnull aResult,
                                         EMError *_Nonnull aError) {
                                       [weakSelf onResult:result
                                           withMethodType:aChannelName
                                                withError:aError
                                               withParams:[aResult
                                                              toJsonObject]];
                                     }];
}

- (void)fetchLastMessageWithChatThreads:(NSDictionary *)param
                         withMethodType:(NSString *)aChannelName
                                 result:(nonnull id<ExtSdkCallbackObjc>)result {
    __weak typeof(self) weakSelf = self;
    NSArray *ids = param[@"threadIds"];
    [EMClient.sharedClient.threadManager
        getLastMessageFromSeverWithChatThreads:ids
                                    completion:^(NSDictionary<NSString *,
                                                              EMChatMessage *>
                                                     *_Nonnull messageMap,
                                                 EMError *_Nonnull aError) {
                                      NSMutableDictionary *ret =
                                          [NSMutableDictionary
                                              dictionaryWithCapacity:0];
                                      for (NSString *key in messageMap
                                               .allKeys) {
                                          EMChatMessage *msg = messageMap[key];
                                          [ret setValue:[msg toJsonObject]
                                                 forKey:key];
                                      }
                                      [weakSelf onResult:result
                                          withMethodType:aChannelName
                                               withError:aError
                                              withParams:ret];
                                    }];
}

- (void)removeMemberFromChatThread:(NSDictionary *)param
                    withMethodType:(NSString *)aChannelName
                            result:(nonnull id<ExtSdkCallbackObjc>)result {
    __weak typeof(self) weakSelf = self;
    NSString *threadId = param[@"threadId"];
    NSString *memberId = param[@"memberId"];
    [EMClient.sharedClient.threadManager
        removeMemberFromChatThread:threadId
                          threadId:memberId
                        completion:^(EMError *_Nonnull aError) {
                          [weakSelf onResult:result
                              withMethodType:aChannelName
                                   withError:aError
                                  withParams:nil];
                        }];
}

- (void)updateChatThreadSubject:(NSDictionary *)param
                 withMethodType:(NSString *)aChannelName
                         result:(nonnull id<ExtSdkCallbackObjc>)result {
    __weak typeof(self) weakSelf = self;
    NSString *threadId = param[@"threadId"];
    NSString *name = param[@"name"];
    [EMClient.sharedClient.threadManager
        updateChatThreadName:name
                    threadId:threadId
                  completion:^(EMError *_Nonnull aError) {
                    [weakSelf onResult:result
                        withMethodType:aChannelName
                             withError:aError
                            withParams:nil];
                  }];
}

- (void)createChatThread:(NSDictionary *)param
          withMethodType:(NSString *)aChannelName
                  result:(nonnull id<ExtSdkCallbackObjc>)result {
    __weak typeof(self) weakSelf = self;
    NSString *name = param[@"name"];
    NSString *messageId = param[@"messageId"];
    NSString *parentId = param[@"parentId"];
    [EMClient.sharedClient.threadManager
        createChatThread:name
               messageId:messageId
                parentId:parentId
              completion:^(EMChatThread *_Nonnull thread,
                           EMError *_Nonnull aError) {
                [weakSelf onResult:result
                    withMethodType:aChannelName
                         withError:aError
                        withParams:[thread toJsonObject]];
              }];
}

- (void)joinChatThread:(NSDictionary *)param
        withMethodType:(NSString *)aChannelName
                result:(nonnull id<ExtSdkCallbackObjc>)result {
    __weak typeof(self) weakSelf = self;
    NSString *threadId = param[@"threadId"];
    [EMClient.sharedClient.threadManager
        joinChatThread:threadId
            completion:^(EMChatThread *_Nonnull thread,
                         EMError *_Nonnull aError) {
              [weakSelf onResult:result
                  withMethodType:aChannelName
                       withError:aError
                      withParams:[thread toJsonObject]];
            }];
}

- (void)leaveChatThread:(NSDictionary *)param
         withMethodType:(NSString *)aChannelName
                 result:(nonnull id<ExtSdkCallbackObjc>)result {
    __weak typeof(self) weakSelf = self;
    NSString *threadId = param[@"threadId"];
    [EMClient.sharedClient.threadManager
        leaveChatThread:threadId
             completion:^(EMError *_Nonnull aError) {
               [weakSelf onResult:result
                   withMethodType:aChannelName
                        withError:aError
                       withParams:nil];
             }];
}

- (void)destroyChatThread:(NSDictionary *)param
           withMethodType:(NSString *)aChannelName
                   result:(nonnull id<ExtSdkCallbackObjc>)result {
    __weak typeof(self) weakSelf = self;
    NSString *threadId = param[@"threadId"];
    [EMClient.sharedClient.threadManager
        destroyChatThread:threadId
               completion:^(EMError *_Nonnull aError) {
                 [weakSelf onResult:result
                     withMethodType:aChannelName
                          withError:aError
                         withParams:nil];
               }];
}

- (void)getChatThread:(NSDictionary *)param
       withMethodType:(NSString *)aChannelName
               result:(nonnull id<ExtSdkCallbackObjc>)result {
    NSString *msgId = param[@"msgId"];
    EMChatMessage *msg =
        [EMClient.sharedClient.chatManager getMessageWithMessageId:msgId];
    if ([self getMessageParams:result
                withMethodType:aChannelName
                   withMessage:msg]) {
        return;
    }
    [self onResult:result
        withMethodType:aChannelName
             withError:nil
            withParams:msg.chatThread != nil ? [msg.chatThread toJsonObject]
                                             : nil];
}

- (void)getThreadConversation:(NSDictionary *)param
               withMethodType:(NSString *)aChannelName
                       result:(nonnull id<ExtSdkCallbackObjc>)result {
    __weak typeof(self) weakSelf = self;
    NSString *conId = param[@"convId"];
    BOOL needCreate = [param[@"createIfNeed"] boolValue];
    EMConversation *conv;
    if (!param[@"isChatThread"]) {
        NSMutableDictionary *newParam =
            [[NSMutableDictionary alloc] initWithDictionary:param];
        newParam[@"isChatThread"] = @(YES);
        EMConversation *conv = [self getConversation:newParam];

    } else {
        EMConversation *conv = [self getConversation:param];
    }
    if (conv) {
        [weakSelf onResult:result
            withMethodType:aChannelName
                 withError:nil
                withParams:[conv toJsonObject]];
    } else {
        EMError *aError = [EMError errorWithDescription:@"no have conversation"
                                                   code:1];
        [weakSelf onResult:result
            withMethodType:aChannelName
                 withError:aError
                withParams:nil];
    }
}

#pragma mark - EMThreadManagerDelegate

- (void)onChatThreadCreate:(EMChatThreadEvent *)event {
    [self onReceive:ExtSdkMethodKeyChatOnChatThreadCreated
         withParams:[event toJsonObject]];
}

- (void)onChatThreadUpdate:(EMChatThreadEvent *)event {
    [self onReceive:ExtSdkMethodKeyChatOnChatThreadUpdated
         withParams:[event toJsonObject]];
}

- (void)onChatThreadDestroy:(EMChatThreadEvent *)event {
    [self onReceive:ExtSdkMethodKeyChatOnChatThreadDestroyed
         withParams:[event toJsonObject]];
}

- (void)onUserKickOutOfChatThread:(EMChatThreadEvent *)event {
    [self onReceive:ExtSdkMethodKeyChatOnChatThreadUserRemoved
         withParams:[event toJsonObject]];
}

@end

File: native_src/java/com/easemob/ext_sdk/dispatch/ExtSdkChatThreadManagerWrapper.java
package com.easemob.ext_sdk.dispatch;

import com.easemob.ext_sdk.common.ExtSdkCallback;
import com.easemob.ext_sdk.common.ExtSdkMethodType;
import com.hyphenate.EMCallBack;
import com.hyphenate.EMChatThreadChangeListener;
import com.hyphenate.EMValueCallBack;
import com.hyphenate.chat.EMChatThread;
import com.hyphenate.chat.EMChatThreadEvent;
import com.hyphenate.chat.EMClient;
import com.hyphenate.chat.EMConversation;
import com.hyphenate.chat.EMCursorResult;
import com.hyphenate.chat.EMMessage;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

public class ExtSdkChatThreadManagerWrapper extends ExtSdkWrapper {
    public static class SingleHolder {
        static ExtSdkChatThreadManagerWrapper instance = new ExtSdkChatThreadManagerWrapper();
    }

    public static ExtSdkChatThreadManagerWrapper getInstance() {
        return ExtSdkChatThreadManagerWrapper.SingleHolder.instance;
    }

    ExtSdkChatThreadManagerWrapper() { registerEaseListener(); }

    private void registerEaseListener() {
        if (this.threadChangeListener != null) {
            EMClient.getInstance().chatThreadManager().removeChatThreadChangeListener(this.threadChangeListener);
        }
        this.threadChangeListener = new EMChatThreadChangeListener() {
            @Override
            public void onChatThreadCreated(EMChatThreadEvent event) {
                ExtSdkWrapper.onReceive(ExtSdkMethodType.onChatThreadCreated,
                                        ExtSdkChatThreadEventHelper.toJson(event));
            }

            @Override
            public void onChatThreadUpdated(EMChatThreadEvent event) {
                ExtSdkWrapper.onReceive(ExtSdkMethodType.onChatThreadUpdated,
                                        ExtSdkChatThreadEventHelper.toJson(event));
            }

            @Override
            public void onChatThreadDestroyed(EMChatThreadEvent event) {
                ExtSdkWrapper.onReceive(ExtSdkMethodType.onChatThreadDestroyed,
                                        ExtSdkChatThreadEventHelper.toJson(event));
            }

            @Override
            public void onChatThreadUserRemoved(EMChatThreadEvent event) {
                ExtSdkWrapper.onReceive(ExtSdkMethodType.onChatThreadUserRemoved,
                                        ExtSdkChatThreadEventHelper.toJson(event));
            }
        };
        EMClient.getInstance().chatThreadManager().addChatThreadChangeListener(this.threadChangeListener);
    }

    public void fetchChatThreadDetail(JSONObject param, String channelName, ExtSdkCallback result)
        throws JSONException {
        String threadId = param.getString("threadId");
        EMClient.getInstance().chatThreadManager().getChatThreadFromServer(
            threadId, new EMValueCallBack<EMChatThread>() {
                @Override
                public void onSuccess(EMChatThread value) {
                    ExtSdkWrapper.onSuccess(result, channelName, ExtSdkChatThreadHelper.toJson(value));
                }

                @Override
                public void onError(int error, String errorMsg) {
                    ExtSdkWrapper.onError(result, error, errorMsg);
                }
            });
    }
    public void fetchJoinedChatThreads(JSONObject param, String channelName, ExtSdkCallback result)
        throws JSONException {
        String cursor = param.getString("cursor");
        int pageSize = param.getInt("pageSize");
        EMClient.getInstance().chatThreadManager().getJoinedChatThreadsFromServer(
            pageSize, cursor, new EMValueCallBack<EMCursorResult<EMChatThread>>() {
                @Override
                public void onSuccess(EMCursorResult<EMChatThread> value) {
                    ExtSdkWrapper.onSuccess(result, channelName, ExtSdkCursorResultHelper.toJson(value));
                }

                @Override
                public void onError(int error, String errorMsg) {
                    ExtSdkWrapper.onError(result, error, errorMsg);
                }
            });
    }
    public void fetchChatThreadsWithParentId(JSONObject param, String channelName, ExtSdkCallback result)
        throws JSONException {
        String parentId = param.getString("parentId");
        String cursor = param.getString("cursor");
        int pageSize = param.getInt("pageSize");
        EMClient.getInstance().chatThreadManager().getChatThreadsFromServer(
            parentId, pageSize, cursor, new EMValueCallBack<EMCursorResult<EMChatThread>>() {
                @Override
                public void onSuccess(EMCursorResult<EMChatThread> value) {
                    ExtSdkWrapper.onSuccess(result, channelName, ExtSdkCursorResultHelper.toJson(value));
                }

                @Override
                public void onError(int error, String errorMsg) {
                    ExtSdkWrapper.onError(result, error, errorMsg);
                }
            });
    }
    public void fetchJoinedChatThreadsWithParentId(JSONObject param, String channelName, ExtSdkCallback result)
        throws JSONException {
        String parentId = param.getString("parentId");
        String cursor = param.getString("cursor");
        int pageSize = param.getInt("pageSize");
        EMClient.getInstance().chatThreadManager().getJoinedChatThreadsFromServer(
            parentId, pageSize, cursor, new EMValueCallBack<EMCursorResult<EMChatThread>>() {
                @Override
                public void onSuccess(EMCursorResult<EMChatThread> value) {
                    ExtSdkWrapper.onSuccess(result, channelName, ExtSdkCursorResultHelper.toJson(value));
                }

                @Override
                public void onError(int error, String errorMsg) {
                    ExtSdkWrapper.onError(result, error, errorMsg);
                }
            });
    }
    public void fetchChatThreadMember(JSONObject param, String channelName, ExtSdkCallback result)
        throws JSONException {
        String threadId = param.getString("threadId");
        String cursor = param.getString("cursor");
        int pageSize = param.getInt("pageSize");
        EMClient.getInstance().chatThreadManager().getChatThreadMembers(
            threadId, pageSize, cursor, new EMValueCallBack<EMCursorResult<String>>() {
                @Override
                public void onSuccess(EMCursorResult<String> value) {
                    ExtSdkWrapper.onSuccess(result, channelName, ExtSdkCursorResultHelper.toJson(value));
                }

                @Override
                public void onError(int error, String errorMsg) {
                    ExtSdkWrapper.onError(result, error, errorMsg);
                }
            });
    }
    public void fetchLastMessageWithChatThreads(JSONObject param, String channelName, ExtSdkCallback result)
        throws JSONException {
        List<String> list = new ArrayList<String>();
        JSONArray array = param.getJSONArray("threadIds");
        for (int i = 0; i < array.length(); i++) {
            list.add(array.getString(i));
        }
        EMClient.getInstance().chatThreadManager().getChatThreadLatestMessage(
            list, new EMValueCallBack<Map<String, EMMessage>>() {
                @Override
                public void onSuccess(Map<String, EMMessage> value) {
                    Map<String, Object> data = new HashMap<>();
                    for (Map.Entry<String, EMMessage> item : value.entrySet()) {
                        data.put(item.getKey(), ExtSdkMessageHelper.toJson(item.getValue()));
                    }
                    ExtSdkWrapper.onSuccess(result, channelName, data);
                }

                @Override
                public void onError(int error, String errorMsg) {
                    ExtSdkWrapper.onError(result, error, errorMsg);
                }
            });
    }
    public void removeMemberFromChatThread(JSONObject param, String channelName, ExtSdkCallback result)
        throws JSONException {
        String threadId = param.getString("threadId");
        String memberId = param.getString("memberId");
        EMClient.getInstance().chatThreadManager().removeMemberFromChatThread(threadId, memberId, new EMCallBack() {
            @Override
            public void onSuccess() {
                ExtSdkWrapper.onSuccess(result, channelName, null);
            }

            @Override
            public void onError(int code, String error) {
                ExtSdkWrapper.onError(result, code, error);
            }
        });
    }
    public void updateChatThreadSubject(JSONObject param, String channelName, ExtSdkCallback result)
        throws JSONException {
        String threadId = param.getString("threadId");
        String name = param.getString("name");
        EMClient.getInstance().chatThreadManager().updateChatThreadName(threadId, name, new EMCallBack() {
            @Override
            public void onSuccess() {
                ExtSdkWrapper.onSuccess(result, channelName, null);
            }

            @Override
            public void onError(int code, String error) {
                ExtSdkWrapper.onError(result, code, error);
            }
        });
    }
    public void createChatThread(JSONObject param, String channelName, ExtSdkCallback result) throws JSONException {
        String name = param.getString("name");
        String messageId = param.getString("messageId");
        String parentId = param.getString("parentId");
        EMClient.getInstance().chatThreadManager().createChatThread(
            parentId, messageId, name, new EMValueCallBack<EMChatThread>() {
                @Override
                public void onSuccess(EMChatThread value) {
                    ExtSdkWrapper.onSuccess(result, channelName, ExtSdkChatThreadHelper.toJson(value));
                }

                @Override
                public void onError(int error, String errorMsg) {
                    ExtSdkWrapper.onError(result, error, errorMsg);
                }
            });
    }
    public void joinChatThread(JSONObject param, String channelName, ExtSdkCallback result) throws JSONException {
        String threadId = param.getString("threadId");
        EMClient.getInstance().chatThreadManager().joinChatThread(threadId, new EMValueCallBack<EMChatThread>() {
            @Override
            public void onSuccess(EMChatThread value) {
                ExtSdkWrapper.onSuccess(result, channelName, ExtSdkChatThreadHelper.toJson(value));
            }

            @Override
            public void onError(int error, String errorMsg) {
                ExtSdkWrapper.onError(result, error, errorMsg);
            }
        });
    }
    public void leaveChatThread(JSONObject param, String channelName, ExtSdkCallback result) throws JSONException {
        String threadId = param.getString("threadId");
        EMClient.getInstance().chatThreadManager().leaveChatThread(threadId, new EMCallBack() {
            @Override
            public void onSuccess() {
                ExtSdkWrapper.onSuccess(result, channelName, null);
            }

            @Override
            public void onError(int code, String error) {
                ExtSdkWrapper.onError(result, code, error);
            }
        });
    }
    public void destroyChatThread(JSONObject param, String channelName, ExtSdkCallback result) throws JSONException {
        String threadId = param.getString("threadId");
        EMClient.getInstance().chatThreadManager().destroyChatThread(threadId, new EMCallBack() {
            @Override
            public void onSuccess() {
                ExtSdkWrapper.onSuccess(result, channelName, null);
            }

            @Override
            public void onError(int code, String error) {
                ExtSdkWrapper.onError(result, code, error);
            }
        });
    }
    public void getMessageThread(JSONObject param, String channelName, ExtSdkCallback result) throws JSONException {
        String msgId = param.getString("msgId");
        EMMessage msg = EMClient.getInstance().chatManager().getMessage(msgId);
        if (ExtSdkWrapper.getMessageParams(msg, channelName, result)) {
            return;
        }
        if (!msg.isChatThreadMessage()) {
            EMChatThread info = msg.getChatThread();
            ExtSdkWrapper.onSuccess(result, channelName, info != null ? ExtSdkChatThreadHelper.toJson(info) : null);
        } else {
            ExtSdkWrapper.onSuccess(result, channelName, null);
        }
    }
    public void getThreadConversation(JSONObject param, String channelName, ExtSdkCallback result)
        throws JSONException {
        if (!param.has("isChatThread")) {
            param.put("isChatThread", true);
        }

        EMConversation conversation = this.getConversation(param);
        ExtSdkWrapper.onSuccess(result, channelName, ExtSdkConversationHelper.toJson(conversation));
    }

    private EMChatThreadChangeListener threadChangeListener = null;
}
